### DISCUSSION TITLE
Interest in adopting a module system similar to cachix/devenv?

### DISCUSSION THREAD

**Discussion by oneingan** [numtide/devshell#316] [Ideas]
I’m curious if there would be interest in adopting a module system similar to the one used in [cachix/devenv](https://github.com/cachix/devenv), instead of the current approach in devshell where features are enabled directly through imports.

Leveraging the module system from devenv could make devshell more modular and user-friendly, allowing us to reuse the extensive knowledge base already present in devenv.

Would this be something the team is open to exploring?

**Comment by zimbatm** [comment]
Yes I think in retrospect that having all the modules pre-imported would make things easier. I wanted to reduce evaluation time, but it won't have that much impact when 20 modules are loaded (compared to the thousants in NixOS).

The big question is how to migrate existing users.

**Comment by oneingan** [comment]
Regarding the migration for existing users, I wanted to share some findings:

The current user base for `extra/services` imports appears to be very low, as indicated by a GitHub search showing no users https://github.com/search?q=lang%3Anix++imports+%22extra%2Fservices%22&type=code.

Similarly, there are only 77 results for `extra/language` imports https://github.com/search?q=lang%3Anix+imports+%22extra%2Flanguage%22&type=code, suggesting limited adoption.

Considering this, the impact of transitioning to a new module system might be minimal. This might simplify the process of migration, given the current scope and usage.

**Reply by zimbatm** [reply]
I think we could also move the modules to a new folder, then shim the original `extra/<name>.nix` modules with `<name>.enable = true`.
